be/src/storage/lake/compaction_policy.cpp

```diff
@@ -532,25 +540,24 @@ StatusOr<CompactionAlgorithm> CompactionPolicy::choose_compaction_algorithm(cons
         ASSIGN_OR_RETURN(auto rowset_iterator_num, rowset->get_read_iterator_num());
         total_iterator_num += rowset_iterator_num;
     }
-    ASSIGN_OR_RETURN(auto tablet_schema, _tablet->get_schema());
-    size_t num_columns = tablet_schema->num_columns();
+    size_t num_columns = _tablet_metadata->schema().column_size();
     return CompactionUtils::choose_compaction_algorithm(num_columns, config::vertical_compaction_max_columns_per_group,
                                                         total_iterator_num);
 }
 
-StatusOr<CompactionPolicyPtr> CompactionPolicy::create_compaction_policy(TabletPtr tablet) {
-    ASSIGN_OR_RETURN(auto tablet_schema, tablet->get_schema());
-    if (tablet_schema->keys_type() == PRIMARY_KEYS) {
-        return std::make_shared<PrimaryCompactionPolicy>(std::move(tablet));
-    }
-    if (config::enable_size_tiered_compaction_strategy) {
-        return std::make_shared<SizeTieredCompactionPolicy>(std::move(tablet));
+StatusOr<CompactionPolicyPtr> CompactionPolicy::create(TabletManager* tablet_mgr,
+                                                       std::shared_ptr<const TabletMetadataPB> tablet_metadata) {
+    if (tablet_metadata->schema().keys_type() == PRIMARY_KEYS) {
+        return std::make_shared<PrimaryCompactionPolicy>(tablet_mgr, std::move(tablet_metadata));
+    } else if (config::enable_size_tiered_compaction_strategy) {
+        return std::make_shared<SizeTieredCompactionPolicy>(tablet_mgr, std::move(tablet_metadata));
+    } else {
+        return std::make_shared<BaseAndCumulativeCompactionPolicy>(tablet_mgr, std::move(tablet_metadata));
     }
-    return std::make_shared<BaseAndCumulativeCompactionPolicy>(std::move(tablet));
 }
 
-double compaction_score(const TabletMetadataPB& metadata) {
-    if (is_primary_key(metadata)) {
+double compaction_score(const std::shared_ptr<const TabletMetadataPB>& metadata) {
```

Comment:
is it still necessary to pass shared_ptr by const reference? or pass by value is good enough?

Reply:
In theory, passing const reference here should still be better than passing by value, as it avoids the extra shared_ptr reference counting increment. Of course, this optimization may be negligible. However, passing const reference here should always be non-negative.